### ISSUE TITLE
jnigen should offer efficient methods to create and extract data from ByteBuffers

### ISSUE THREAD

**Issue by brianquinlan** [dart-lang/jnigen#387]
For example:
```dart
// Get the data out of a byte buffer without having to index into it's array 1,000,000 times.
final b = java.ByteBuffer.allocateDirect(1000000);
UInt8List d = b.toUInt8List();

// Populate a ByteBuffer without having to index into an array 1,000,000 times
final d = UInt8List(1000000);
java.ByteBuffer b = d.toJByteBuffer();
```

The JNI functions that can help with this are:
- [NewDirectByteBuffer](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#NewDirectByteBuffer)
- [GetDirectBufferAddress](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetDirectBufferAddress)
- [GetDirectBufferCapacity](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetDirectBufferCapacity)

**Comment by brianquinlan** [comment]
I'm creating bindings for Cronet and have two use cases:

1.

I need to convert a complete `java.nio.ByteBuffer` into a `List<int>`. The code currently looks like:

```dart
      onReadCompleted: (urlRequest, responseInfo, byteBuffer) {
        byteBuffer.flip();

        final remaining = byteBuffer.remaining();
        final data = Uint8List(remaining);
        // TODO: Use a more efficient approach when
        // https://github.com/dart-lang/jnigen/issues/387 is fixed.
        for (var i = 0; i < remaining; ++i) {
          data[i] = byteBuffer.get1(i);
        }
        responseStream!.add(data);
        byteBuffer.clear();
        cronetRequest.read(byteBuffer);
      },
```

2.

I have a `List<int>` and need to convert it to a `java.nio.ByteBuffer` (or `byte[]`). The code currently looks like:

```dart
    if (body.isNotEmpty) {
      // TODO: Use a more efficient approach when
      // https://github.com/dart-lang/jnigen/issues/387 is fixed.
      final bodyBytes = JArray(jbyte.type, body.length);
      for (var i = 0; i < body.length; ++i) {
        bodyBytes[i] = body[i];
      }
      builder.setUploadDataProvider(
          // This also accepts a `java.nio.ByteBuffer`
          jb.UploadDataProviders.create4(bodyBytes), _executor);
```

Eventually I'll want to implement by own `UploadDataProvider`, which will mean copying the data from a `List<int>` to an offset in a `java.nio.ByteBuffer`

**Comment by HosseinYousefi** [comment]
Fixed by #390.
